src/async_impl/decoder.rs

```diff
@@ -0,0 +1,297 @@
+/*!
+A potentially non-blocking response decoder.
+
+The decoder wraps a stream of chunks in a `Read` interface. 
+If the underlying chunk stream never blocks (by returning `Err(WouldBlock)`),
+then the decoder will also never return `Err(WouldBlock)`.
+This is essential for it to be usable in synchronous scenarios.
+
+If the response is potentially gzipped, then we need to read a byte to make sure the response isn't actually empty.
+This is also done in an asynchronous in case the response body isn't immediately available.
+
+This module consists of a few main types:
+
+- `ReadableChunks` is a `Read`-like wrapper around a stream
+- `PollChunk` is a stream of chunks, that may or may not be non-blocking
+- `Decoder` is a layer over `ReadableChunks` that applies the right decompression
+
+The following types directly support the gzip compression case:
+
+- `DecoderInnerFuture` is a non-blocking constructor for a `Decoder` in case the body needs to be read
+- `Peeked` is a wrapper around a peeked byte that prevents an EOF on read
```

Comment:
This is because `libflate` does a `read_exact` for 2 bytes, which will return an eof error if there aren't 2 bytes available.

So we've got a wrapper that reads a byte and checks whether the response is actually empty.